### ISSUE TITLE
Feature+Docs: PodSecurityPolicy for Windows

### ISSUE THREAD

**Issue by PatrickLang** [kubernetes/kubernetes#64801]
Most of the fields in [PodSecurityPolicy](https://kubernetes.io/docs/concepts/policy/pod-security-policy/) today cannot be implemented in Windows since they're based on Linux namespace and UID/GID assumptions.

We need a feature proposal to add or modify what can be implemented on Windows to achieve use cases such as:

- Restricting host path mounts (allowedFlexVolumes, allowedHostPaths)
- Restricting runAsUser based on username[string] or group membership rather than uid/gid
- Using group managed service accounts (related to https://github.com/kubernetes/kubernetes/issues/62038)

Areas that need to be documented as does not apply to Windows:

- hostPID, hostIPC, hostNetwork, hostPorts - these are not possible today without changes to Windows itself, could be considered for future
- fsGroup
- *Capabilities, seLinux, AppArmor, sysctl - these are all Linux specific

**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**Environment**:
- Kubernetes version (use `kubectl version`): v1.9, v1.10, v1.11
- Cloud provider or hardware configuration: not specific
- OS (e.g. from /etc/os-release): Windows Server 2016, Windows Server version 1709, Windows Server version 1803

**Comment by tallclair** [comment]
I'd like to redo PodSecurityPolicy at some point, and don't necessarily see it progressing to GA in its current form. Given that, I think it will be useful to understand the windows requirements for PodSecurityPolicy to inform the design of the successor, but I don't think it's worth investing in retrofitting the current PSP for windows.

**Comment by feiskyer** [comment]
> I'd like to redo PodSecurityPolicy at some point, and don't necessarily see it progressing to GA in its current form. 

Any timeline for this?

**Comment by tallclair** [comment]
No. Discussions are underway ([wg-policy](https://github.com/kubernetes/community/tree/master/wg-policy)) about what should be a native policy vs. a policy DSL (e.g. OPA/rego). Any progress on PodSecurityPolicy is blocked on that guidance.

**Comment by derekwaynecarr** [comment]
@PatrickLang we need to clarify the pod spec for windows before making updates around the policy discussion.

**Comment by nileshkumar01** [comment]
@PatrickLang This is something that the client i am working with is also very interested in getting working. I tested the volume mount policy for Windows and it doesn't seem to be working.

**Comment by ddebroy** [comment]
@nileshkumar01 can you post the exact policy YAML you tried that did not work?

**Comment by tallclair** [comment]
Copied from [mailing list](https://groups.google.com/d/topic/kubernetes-sig-windows/axUqMtB13fU/discussion) since it's relevant to this issue:

I'm working on putting together SIG-Auth recommended policy "profiles" for securing pods. You can think of these as predefined PodSecurityPolicies, but the final product will be an implementation-agnostic spec.

The 3 profiles we're currently planning are:

Privileged - Totally unrestricted, any pod options can be specified.
Default - The pod default options are allowed (e.g. run as root, CAP_NET_RAW, unconfined seccomp), more restrictive options are allowed, but less restrictive options are forbidden (e.g. HostNetwork). Strikes a balance between best practices and compatibility.
Restricted - The currently recommended set of best practices (e.g. run as non-root) are required.

I know windows containers can't share host namespaces, but are there options to escalate the privileges of windows pods? Or would the privileged and default profiles be essentially equivalent for windows containers?

What about the best-practices for windows containers? Are there generic hardening recommendations beyond the defaults?

**Comment by nileshkumar01** [comment]
> @nileshkumar01 can you post the exact policy YAML you tried that did not work?

Hi @ddebroy i created my dev environment and was able to get the volume mount policy to work.

**Comment by jingxu97** [comment]
anyone know whether allowedHostPaths works for windows?

**Comment by jsturtevant** [comment]
a few related issues that are being worked on related to MustRunAsNonRoot:

- #91482 
- #91414

and the pr to fix: https://github.com/kubernetes/kubernetes/pull/92355

**Comment by jsturtevant** [comment]
This has come up quite a bit and starting to find more scenarios where although it should be ignore not affect windows pods they are either blocked or don't get scheduled. I will start to look into more as well as the how it might work in regards to Gatekeeper.

/assign

**Comment by tallclair** [comment]
A good starting point might be to document that pod features that should be restricted on windows, as part of the ["Pod Security Standards"](https://kubernetes.io/docs/concepts/security/pod-security-standards/)

**Comment by jsturtevant** [comment]
I went through the Pod Security Standards doc and mapped each to what can/should apply to windows and things that don't make sense. If folks want to have a look and leave comments we can discuss at a sig-windows meeting: https://docs.google.com/spreadsheets/d/1Snhqtnu38I2Sqa7UQQbD3ZwYTkXMo7L81lUHSmh7buU/edit?usp=sharing

For areas where the security context was causing issues with PSP these PR's should take care of them: https://github.com/kubernetes/kubernetes/pull/93475 and https://github.com/kubernetes/kubernetes/issues/91482

In summary the following should be available in windows and need to map to a particular "standard".  

- Usage of volume types
- Usage of the host filesystem
- Running as Non-root

There are on windows specific one that should be added: 

- GMSA

**Comment by jsturtevant** [comment]
With the proposal for the [Privileged Containers KEP](https://docs.google.com/document/d/12EUtMdWFxhTCfFrqhlBGWV70MkZZPOgxw0X-LTR0VAo/edit#), there was an analysis done for what would be potentially impacted in regards to Pod Security Standards and policies.  Please see the [KEP for details](https://docs.google.com/document/d/12EUtMdWFxhTCfFrqhlBGWV70MkZZPOgxw0X-LTR0VAo/edit#heading=h.lfmx6g5l0bof).

**Comment by jglick** [comment]
Can this be closed given #97171?

**Comment by liggitt** [comment]
for PodSecurityPolicy specifically, I don't expect to see any windows-related changes, so this issue could probably be closed, but the points raised here (how OS-specific implications shape policy) would be good for the group working on a proposal for a PSP replacement to consider/incorporate if possible

cc @IanColdwater @tabbysable @tallclair

**Comment by jsturtevant** [comment]
agree for PSP but would like to get update guidance for general Windows Pod Security Standards doc: https://kubernetes.io/docs/concepts/security/pod-security-standards/

**Comment by tabbysable** [comment]
@jsturtevant It sounds like y'all in SIG Windows know what the hardening guidelines you have in mind are (at least somewhat, with the mappings you've done already), but would like some additional eyes on it and assistance with writing a PR to the Pod Security Standards doc? I'd recommend asking about that in the sig-security-docs channel on kubernetes slack to see if anybody has bandwidth to help.

**Comment by jsturtevant** [comment]
Before writing the docs I wanted to ensure we had these working in some form (most likely via Gatekeeper) so if we publish guidance it is something that can be followed and implemented. I had to step away from this shortly after doing the initial analysis for other work. Hoping I can get back to it soon.

**Comment by tabbysable** [comment]
Awesome! I assume you'll be PRing an example policy into https://github.com/open-policy-agent/gatekeeper-library/. Feel free to drop into #sig-security on k8s slack to ask if anyone wants to review it!

**Comment by jsturtevant** [comment]
We got the first gate-keeper policy PR open for anyone following along: https://github.com/open-policy-agent/gatekeeper-library/pull/70

**Comment by tallclair** [comment]
I think we can close this, since PSP is marked as deprecated and targeting removal in v1.25
